### DISCUSSION TITLE
zoau 1.3.0.0 dataset create error - IndexError: list index out of range

### DISCUSSION THREAD

**Discussion by gngrossi** [IBM/zoau-samples#73] [Q&A]
Recently installed Python 3.12 and zoau 1.3.0.0 on z/OS 2.5
Receiving errors trying to create any dataset type although the allocation is successful.
[zoau1300_errors.txt](https://github.com/IBM/zoau-samples/files/13981116/zoau1300_errors.txt)

**Comment by AnthonyGiorgio** [comment]
Make sure you update your python language bindings to the version that comes with v1.3.0, as well as updating your python code to use the new APIs.  There are significant changes with the dataset and jobs classes, [so consult the migration guide for details](https://www.ibm.com/docs/en/zoau/1.3.x?topic=planning-migrating-zoau-v13-from-v12x-earlier).

**Comment by gngrossi** [comment]
Testing zoau 1.3.0.0 with Python 3.12.

bash-5.2$ python3 --version
Python 3.12.0
bash-5.2$ pip list | grep -E "Package|zoau"
Package         Version
zoautil-py      1.3.0.0

Encountering several issues with datasets:
1) create will allocate a dataset but then errors
2) _create is successful using the same parameters
3) _create can't create VSAM (KSDS, ESDS, RRDS) files
3) exists returns False when the dataset is allocated (and contains data)
4) copy or _copy doesn't support PDS or PDSE

--

Python 3.12.0 (heads/pyz_dev-3.12:ef647e3673, Oct 31 2023, 19:02:59) [Clang 14.0.0 (build 1465bdb)] on zos
Type "help", "copyright", "credits" or "license" for more information.
>>> from zoautil_py import datasets
>>> print(dir(datasets))
['Dataset', 'DatasetFetchException', 'DatasetVerificationError', 'DatasetWriteException', 'MissingFunctionParameter', 'Optional', 'Union', 'ZOAUException', 'ZOAUResponse', '__annotations__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_compare', '_copy', '_create', '_delete', '_delete_members', '_dls_call', '_dunzip', '_dzip', '_find_member', '_find_replace', '_get_hlq', '_list_datasets_by_volume', '_list_members', '_list_vsam_datasets', '_move', '_move_member', '_read', '_read_as_bytes', '_read_head', '_search', '_tmp_name', '_write', 'blockinfile', 'call_zoau_library', 'call_zoau_library_bin', 'cast_int_optional', 'clean_shell_input', 'compare', 'copy', 'create', 'datetime', 'delete', 'delete_members', 'dunzip', 'dzip', 'escape_block_string', 'exists', 'find_member', 'find_replace', 'get_hlq', 'lineinfile', 'list_dataset_names', 'list_datasets', 'list_datasets_by_volume', 'list_members', 'list_vsam_datasets', 'logger', 'logging', 'move', 'move_member', 'parse_universal_arguments', 'parse_unknown', 'read', 'read_as_bytes', 'read_head', 'search', 'tmp_name', 'warnings', 'write', 'zoau_json_load']
>>>

**Comment by AnthonyGiorgio** [comment]
Could you provide us with the [ZOAU debug output](https://www.ibm.com/docs/en/zoau/1.3.x?topic=troubleshooting#tips-for-the-python-api)?

**Comment by mnchava** [comment]
Could you please include more information on the code that is failing? It would be helpful to see the response of `_create` and `_dls_call` (or `_listvsam_datasets` if you are also getting errors with VSAMs). You can also enable debug logging to get some more information on the ZOAU commands being used: https://www.ibm.com/docs/en/zoau/1.3.x?topic=troubleshooting#tips-for-the-python-api

Unfortunately, `exists()` only works for non-VSAM datasets right now, are you trying to check VSAMs?

What errors do you get when trying to copy partitioned datasets? `copy` does support PDS and PDSE, just specify the member name in the source argument: `datasets.copy('DATA.SAMPLE(A)', 'DATA.SAMPLE(B)')`

**Comment by gngrossi** [comment]
Example of a copy of a pds and pdse
The copy of the pds looked liked it worked but the dataset wasn't cataloged (provided a TSO/ISPF DSLIST)
[zoau13_copy_pds_pdse.txt](https://github.com/IBM/zoau-samples/files/14030606/zoau13_copy_pds_pdse.txt)

**Comment by gngrossi** [comment]
Example of a create and  _create of a sequential dataset
The create errored and _create completed
[zoau13_create_seq.txt](https://github.com/IBM/zoau-samples/files/14032780/zoau13_create_seq.txt)

**Comment by gngrossi** [comment]
Example of a _create of a VSAM KSDS dataset
The create looked like it worked but the dataset wasn't cataloged (provided a TSO/ISPF DSLIST)
[zoau13_create_ksds.txt](https://github.com/IBM/zoau-samples/files/14032863/zoau13_create_ksds.txt)

**Comment by gngrossi** [comment]
Example of a creating a sequential dataset and checking if it exists
[zoau13_exists_seq.txt](https://github.com/IBM/zoau-samples/files/14032954/zoau13_exists_seq.txt)

**Comment by gngrossi** [comment]
Example of list datasets
list_datasets and list_dataset_names return empty and list_vsam_datasets errors (provided a TSO/ISPF DSLIST snippet)
[zoau13_list_datasets.txt](https://github.com/IBM/zoau-samples/files/14039979/zoau13_list_datasets.txt)

**Comment by gngrossi** [comment]
Using the os interface for tsocmd seems to work well.
>>> import sys
>>> stream = os.popen('/bin/tsocmd "delete t.from.python"')
>>> delete t.from.python

>>> stream = os.popen('/bin/tsocmd "alloc dsname(t.from.python) space(10,1) tracks recorg(es) lrecl(80) blksize(8192) new"')
>>> alloc dsname(t.from.python) space(10,1) tracks recorg(es) lrecl(80) blksize(8192) new

>>> stream = os.popen('tsocmd "listc lvl(''@02858.t.from'')"')
>>> listc lvl(@02858.t.from)
>>> output = stream.read()
>>> print(output)
CLUSTER ------- @02858.T.FROM.PYTHON
     IN-CAT --- SYS1.UCATTSO
DATA ---------- @02858.T.FROM.PYTHON.DATA
     IN-CAT --- SYS1.UCATTSO

**Comment by ehrocha** [comment]
Hello All, 
I was having the exact same issue, the problem with my installation was the filesystem was not mounted with the setuid parameter. 
This is the error message I was getting: 
```
    "msg": "DatasetCreateError('An error occurred during creation of data set \"A089712.P7174511.T0635200.C0000000\". Unable to verify the data set was created. Received DatasetVerificationError from ZOAU.')",
    "ret_code": {
        "code": 8
```

The problem was revealed with the "zoaversion -c" command for diagnostics:
```
Validating ZOAU installation in ZOAU_HOME
Error: ZOAU filesystem at /usr/lpp/IBM/zoau/v1r3 is mounted with the NOSETUID option.
ZOAU will not work correctly in this configuration.
Remount the filesystem without the NOSETUID option.
```
So I have changed my automount policy to mount the file with setuid (the same can be done with BPXPRMxx):
```
# =====================================
# Mounts for zoau
# =====================================
 name       zoau13
 type       ZFS
 filesystem SYS9.ZOAU13.OMVS.SBGYROOT
 mode       rdwr
 duration   60
 delay      10
 setuid     yes
```
After applying the new automount policy and remounting the filesystem problem was cleared.

**Reply by AnthonyGiorgio** [reply]
I'm positively **thrilled** that my little diagnostics script helped you out.  Thank you for the feedback!
